bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java

```diff
@@ -133,17 +133,31 @@ public void stop()
 
         List<Object> reversedInstances = new ArrayList<>(managedInstances);
         Collections.reverse(reversedInstances);
+        LifeCycleStopException stopException = null;
 
         for (Object obj : reversedInstances) {
             log.debug("Stopping %s", obj.getClass().getName());
             LifeCycleMethods methods = methodsMap.get(obj.getClass());
             for (Method preDestroy : methods.methodsFor(PreDestroy.class)) {
                 log.debug("\t%s()", preDestroy.getName());
-                preDestroy.invoke(obj);
+                try {
+                    preDestroy.invoke(obj);
+                }
+                catch (Exception e) {
+                    Throwable t = (e instanceof InvocationTargetException) ? ((InvocationTargetException) e).getTargetException() : e;
+                    log.error("Exception in PreDestroy method %s::%s()", obj.getClass().getName(), preDestroy.getName(), t);
```

Comment:
Is `t` being logged? (fmt string has 2 `%s`s only)

If you intent to log the stacktrace in full, `t` should be the first param... but I don't think it's a good idea, since `t` gets propagated eventually.

Reply:
@electrum if there are multiple components failing, logging one line immediately can improve log output readability. 
IMO it's better to log here, but just `t` `toString()` instead of full stacktrace.

Reply:
The general rule is "log or throw, not both".

Reply:
Re: the first comment, I didn't look hard enough at the logger API used here. Throwable should indeed be the first argument if being logged. Will do that.

Regarding the log ^ throw debate, I'm torn about how to proceed. If the lifecycle is stopping because the JVM is exiting, it definitely makes sense to log errors immediately because the logger infrastructure may be shutting down or the JVM may force-exit after some timeout. Logging immediately should be clearly preferable.

However, lifecylce stops initiated by an application that will continue to run will probably prefer to handle the exceptions all at once, probably via a checked exception type so there's only one type of exception to handle the problematic (or "dirty") shutdown.

These two use cases seem at odds with one another, so we can either:
- pick one at the detriment of the other use case
- break the log ^ throw convention in this particular case
- make these two use cases call distinct methods and support both modes.

I'm going to pursue the two method approach and update the CR, but I'm happy to implement either of the other two options if you guys prefer. Let me know.